File: Python/0040_combination_sum_ii/0040_combination_sum_ii.py
class Solution:
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        def backtrack(candidates, target, i, current):
            if target < 0 or i > len(candidates):
                return
            if target == 0:
                res.append(current)
                return
            if i <= len(candidates) - 1:
                backtrack(candidates, target - candidates[i], i + 1, current + [candidates[i]])
                backtrack(candidates, target, i + 1, current)

        res = []
        candidates = sorted(candidates)
        backtrack(candidates, target, 0, [])
        no_duplicate_res = []
        for i in res:
            if i not in no_duplicate_res:
                no_duplicate_res.append(i)        
        return no_duplicate_res

File: Java/0040_combination_sum_ii/0040_combination_sum_ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(candidates);
        backtrack(res, new ArrayList<Integer>(), candidates, target, 0);
        res = res.stream().distinct().collect(Collectors.toList());
        return res;
    }

    private void backtrack(List<List<Integer>> res, ArrayList<Integer> current, int[] candidates, int target, int index) {
        if (target < 0 || index > candidates.length)
            return;
        if (target == 0) {
            res.add(new ArrayList<Integer>(current));
            return;
        }
        else if (index <= candidates.length - 1) {
            current.add(candidates[index]);
            backtrack(res, current, candidates, target - candidates[index], index + 1);
            current.remove(current.indexOf(candidates[index]));
            backtrack(res, current, candidates, target, index + 1);
        }
    }
}